experiments/targeting.go

```diff
@@ -0,0 +1,357 @@
+package experiments
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+// Targeting is the common interface to implement experiment targeting.
+// Evaluated whether the provided input matches the expected values.
+type Targeting interface {
+	Evaluate(inputs map[string]interface{}) bool
+}
+
+// NewTargeting parses the given targeting configuration into a Targeting.
+func NewTargeting(targetingConfig []byte) (Targeting, error) {
+	var config map[string]interface{}
+	decoder := json.NewDecoder(bytes.NewBuffer(targetingConfig))
+	// ensures numbers are parsed into json.Number in order to differentiate
+	// between integer and float64 later on
+	decoder.UseNumber()
+	err := decoder.Decode(&config)
+	if err != nil {
+		return nil, err
+	}
+	if len(config) != 1 {
+		return nil, TargetingNodeError("call to create targeting tree expects single input key")
+	}
+	return parseNode(config)
+}
+
+func parseNode(node interface{}) (Targeting, error) {
+	switch n := node.(type) {
+	case map[interface{}]interface{}:
```

Comment:
`map[interface{}]interface{}` is encountered during parsing for the lower level JSON objects. The first level is detected by the JSONUnmarshaller as `map[string]interface{}`, every subsequent `map[interface{}][interface{}`.

Reply:
do you have a concrete example? I think here we always try to convert the key to string anyways, which kind of implies that the key is supposed to be string? I also don't think I've seen a map in json that the key is not string.

Reply:
You're right about that, that is the only case. I am just saying, that when the JSON document gets unmarshalled by Go into `map[string]interface{}]` then every value of that map is typed `map[interface{}][interface{}]`, because that's what the default JSON unmarshaller does even though every key is indeed a string. We can probably work around that behavior but that's just me dealing with what I received.